runtime/near-vm-runner/src/logic/logic.rs

```diff
@@ -2794,15 +2768,27 @@ impl<'a> VMLogic<'a> {
     /// distributed to functions that specify a gas weight. If there are no functions with
     /// a gas weight, the outcome will contain unused gas as usual.
     pub fn compute_outcome_and_distribute_gas(mut self) -> VMOutcome {
-        if !self.context.is_view() {
-            // Distribute unused gas to scheduled function calls
+        'distribute_gas: {
+            if self.context.is_view() {
+                break 'distribute_gas;
+            }
+            let mut gas_weight_sum: u128 = 0;
+            self.ext
+                .unused_gas_recipients(&mut |_, weight, _| gas_weight_sum += u128::from(weight.0));
             let unused_gas = self.gas_counter.unused_gas();
-
-            // Spend all remaining gas by distributing it among function calls that specify
-            // a gas weight
-            if let GasDistribution::All = self.receipt_manager.distribute_unused_gas(unused_gas) {
-                self.gas_counter.prepay_gas(unused_gas).unwrap();
+            if gas_weight_sum == 0 || unused_gas == 0 {
+                break 'distribute_gas;
             }
+            let mut distributed = 0;
+            self.ext.unused_gas_recipients(&mut |gas, weight, is_last| {
+                let assign = (unused_gas as u128 * weight.0 as u128 / gas_weight_sum) as u64;
+                *gas += assign;
+                distributed += assign;
+                if is_last {
+                    *gas += unused_gas - distributed;
+                }
+            });
+            self.gas_counter.prepay_gas(unused_gas).unwrap();
```

Comment:
I wonder, isn't this a bit too fancy for the job? Shouldn't we just move all the logic out of here?

I saw your comment that you want to make gas distribution logic something done by the VM runtime, while handling the outgoing receipts is more of a transaction runtime thing. But seeing the tricks necessary to make this cut, I find it hart to believe it is a clean cut.

Another way of looking at it would be to say the `VMOutcome` should just contain the gas weight and leave it up to the transaction runtime to apply them. This isn't burning any gas, hence it doesn't affect gas profiles. And it certainly doesn't affect the main outcome of the function call, only the newly spawned receipts. With that view, I think gas distribution can be seen as a transaction runtime thing, not a VM thing.

WDYT, would that makes sense? I'm not convinced in either direction yet (distributing in VM or in tx runtime) but in my mind, moving all to the transaction runtime would remove a chunk of complicated code.

Reply:
Sorry for the delay!

So I tried to look at those tests, I only found `function_call_weight_basic_cases_test` and `function_call_no_weight_refund`. The former calls `function_call_weight_check` 10 times, so there are 10 test cases, but the code only needs to be changes in ``function_call_weight_check`.

I believe all other calls to `compute_outcome_and_distribute_gas` don't test anything related to gas weights. They just want to produce a `VMOutcome` but they don't test anything specific to gas weights.

So have to admit I struggle to understand the WWII scenario you faced.^^ Can you maybe elaborate a bit more?


> > VMOutcome should just contain the gas weight
> 
> In a design where the transaction runtime is responsible for dealing with gas weights, `VMOutcome` or really the contract runtime as a whole, needn’t to think about weights at all, other than passing through information about the queued function call actions (which it already does.)

Ahh, yes you are right. Make me think even more that we should move this logic out of the VM.

Reply:
> Make me think even more that we should move this logic out of the VM.

I guess it doesn’t hurt to try again :shrug: 

> So have to admit I struggle to understand the WWII scenario you faced.^^ Can you maybe elaborate a bit more?

I’m quite sure there were more tests than these whose output was dependent on the gas weights, but I can revisit those :shrug:

Reply:
So, there are two tests that are failing after the most recent changes:

* `attach_unspent_gas_but_burn_all_gas`; and
* `function_call_weight_basic_cases_test`.

I only looked at moving the former to `node-runtime` today. The unfortunate thing about the test being there is that none of the test frameworks (e.g. `RuntimeGroup`) work at a low enough level to enable peeking into the promises/actions generated by the contract. The only two viable entry-points for this test from the transaction runtime are `actions::action_function_call` or `actions::execute_function_call`. These are pretty painful to set up an invocation environment for T_T

Looking at `function_call_weight_basic_cases_test` it is using similarly "low-level" APIs to check for what it is checking. To do the same with `actions::execute_function_call` would require some serious rethinking of how these tests work. But I think there's a more straightforward solution – I could move the gas distribution logic entirely into `ReceiptManager` and have this test as part of its unit tests.

Reply:
Thanks for elaborating!

> I could move the gas distribution logic entirely into ReceiptManager and have this test as part of its unit tests.

That sounds quite reasonable to me.

But overall, since you have already spent quite some time, let me ask this: Do you think it's worth going that route? Or would you rather merge the original proposal, where gas distribution is split between VM and transaction runtime?
Obviously, since you are much deeper into the code right now, I would trust your judgement much more than my own.

Reply:
My intuition is that it is correct to have this code as part of the transaction runtime, and I suspect that we’re better off removing the test(s) that cannot be expressed rather than keeping the logic itself in a less than ideal state and keep the tests.

Reply:
ok, sounds good to me